picus/algorithms/dpvl.rkt

```diff
@@ -303,39 +302,61 @@
 )
 
 ; this creates a new hash with r1cs variables replaced by corresponding circom variables
-; (note) this will remove helping variables like "one", "ps?", etc.
-(define (map-to-vars info path-sym)
-    (define rd (csv->list (open-input-file path-sym)))
-    ; create r1cs-id to circom-var mapping
-    (define r2c-map (make-hash (for/list ([p rd])
-        (cons (list-ref p 0) (list-ref p 3))
-    )))
-    (define pinfo (if (list? info) (make-hash) info)) ; patch for info type, fix later
+; (note) when unmappable? is #f, this will remove helping variables like "one", "ps?", etc.
+;; map-to-vars :: (listof hash?), path-string?, #:unmappable? boolean? -> (listof hash?)
+(define (map-to-vars info path-sym #:unmappable? [unmappable? #f])
+  (define rd (csv->list (open-input-file path-sym)))
+  ; create r1cs-id to circom-var mapping
+  (define r2c-map
+    (make-hash (for/list ([p rd]) (cons (list-ref p 0) (list-ref p 3)))))
+  (define (process-subinfo pinfo)
     (define new-info (make-hash))
-    (for ([k (hash-keys pinfo)])
-        (cond
-            [(equal? k "x0") (void)] ; skip since this is a constant
-            [(string-prefix? k "x")
-                (define rid (substring k 1))
-                (define cid (hash-ref r2c-map rid))
-                (define val (hash-ref pinfo k))
-
-                (define sid (format "m1.~a" cid))
-                (hash-set! new-info sid val)
-            ]
-            [(string-prefix? k "y")
-                (define rid (substring k 1))
-                (define cid (hash-ref r2c-map rid))
-                (define val (hash-ref pinfo k))
-
-                (define sid (format "m2.~a" cid))
-                (hash-set! new-info sid val)
-            ]
-            [else (void)] ; skip otherwise
-        )
-    )
-    new-info
-)
+    (for ([(k val) (in-hash pinfo)])
+      (cond
+        [(string-prefix? k "x")
+         (define rid (substring k 1))
+         (define cid (hash-ref r2c-map rid))
+
+         (define sid (format "m1.~a" cid))
+         (hash-set! new-info sid val)]
+        [(string-prefix? k "y")
+         (define rid (substring k 1))
+         (define cid (hash-ref r2c-map rid))
+
+         (define sid (format "m2.~a" cid))
+         (hash-set! new-info sid val)]
+        [else
+         (when unmappable?
+           (hash-set! new-info k val))]))
+    new-info)
+  (map process-subinfo info))
+
+;; partition-vars :: (or/c '() hash?), set?, set? -> (list/c hash? hash? hash? hash?)
+(define (partition-vars info input-set output-set)
+  (define pinfo (if (list? info) (make-hash) info)) ; patch for info type, fix later
+  (for/fold ([input-info (hash)]
+             [output1-info (hash)]
+             [output2-info (hash)]
+             [other-info (hash)]
+             #:result (list input-info output1-info output2-info other-info))
+            ([(k v) (in-hash pinfo)])
+    (match k
+      [(pregexp #px"^x(\\d+)$" (list _ (app string->number n)))
```

Comment:
NIT: Let's add a comment explaining all regexes we use.

Reply:
Fixed.